BasicBackwardsCompatibilityIT.testIndexRollingUpgrade sometimes fails

Example failure:
http://build-us-00.elastic.co/job/es_core_2x_window-2008/712

Reproduce command (I can't reproduce this locally):

```
REPRODUCE WITH: mvn verify -Pdev -Dskip.unit.tests -pl org.elasticsearch.qa.backwards:2.0 -Dtests.seed=FE8754AE850E4A36 -Dtests.class=org.elasticsearch.bwcompat.BasicBackwardsCompatibilityIT -Dtests.method="testIndexRollingUpgrade" -Des.logger.level=DEBUG -Dtests.nightly=false -Dtests.client.ratio=0.0 -Dtests.heap.size=512m -Des.node.mode=local -Dtests.jvm.argline="-server -XX:+UseG1GC -XX:+UseCompressedOops -Djava.net.preferIPv4Stack=true" -Dtests.locale=sr-BA -Dtests.timezone=Australia/Tasmania
```

Failure:

```
FAILURE 11.8s | BasicBackwardsCompatibilityIT.testIndexRollingUpgrade <<<
   > Throwable #1: java.lang.AssertionError: Count is 96 but 116 was expected.  Total shards: 14 Successful shards: 14 & 0 shard failures:
   > 	at __randomizedtesting.SeedInfo.seed([FE8754AE850E4A36:88FA0EA86CDC10E0]:0)
   > 	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:244)
   > 	at org.elasticsearch.bwcompat.BasicBackwardsCompatibilityIT.testIndexRollingUpgrade(BasicBackwardsCompatibilityIT.java:371)
   > 	at java.lang.Thread.run(Thread.java:745)
```

--- a/qa/backwards/shared/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java
+++ b/qa/backwards/shared/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java
@@ -353,6 +353,9 @@ public class BasicBackwardsCompatibilityIT extends ESBackcompatTestCase {
             CountResponse countResponse = client().prepareCount().get();
             assertHitCount(countResponse, numDocs);
             assertSimpleSort("num_double", "num_int");
+            /* There is a bug where we do not fsync the translog after recovery from primary which causes dataloss. It is fixed in all
+            versions >=2.1.2. But there was no 2.0.3 release which would have the fix for 2.0. Instead, we have to flush here so we do not
+            run into this bug in the test. See https://github.com/elastic/elasticsearch/pull/15832 */
             if (compatibilityVersion().onOrBefore(Version.V_2_0_2)) {
                 flush();
             }

--- a/qa/backwards/shared/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java
+++ b/qa/backwards/shared/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java
@@ -353,6 +353,9 @@ public class BasicBackwardsCompatibilityIT extends ESBackcompatTestCase {
             CountResponse countResponse = client().prepareCount().get();
             assertHitCount(countResponse, numDocs);
             assertSimpleSort("num_double", "num_int");
+            if (compatibilityVersion().onOrBefore(Version.V_2_0_2)) {
+                flush();
+            }
             upgraded = backwardsCluster().upgradeOneNode();
             ensureYellow();
             countResponse = client().prepareCount().get();

